contentcuration/contentcuration/utils/celery/tasks.py

```diff
@@ -0,0 +1,141 @@
+import math
+from functools import partial
+
+from celery import states
+from celery.app.task import Task
+from celery.result import AsyncResult
+
+from contentcuration.utils.sentry import report_exception
+
+
+class ProgressTracker:
+    """
+    Helper to track task progress
+    """
+    key = 'progress'
+
+    def __init__(self, task_id, update_state):
+        """
+        :param task_id: The ID of the calling task
+        :param update_state: Callback to update the task's state
+        :type update_state: Callable
+        """
+        self.task_id = task_id
+        self.update_state = update_state
+        self.total = 100.0
+        self.progress = 0.0
+        self.last_reported_progress = 0.0
+
+    def set_total(self, total):
+        """
+        :param total: The amount for which progress is deemed 100%, defaults to 100
+        """
+        self.total = total
+
+    def increment(self, increment=1.0):
+        """
+        Increments and triggers tracking of the progress to the task meta
+        """
+        self.track(self.progress + increment)
+
+    def track(self, progress):
+        """
+        :param progress: The current progress amount
+        """
+        self.progress = progress
+
+        # only update the task progress in >=1% increments
+        if math.floor(self.last_reported_progress) < math.floor(self.task_progress):
+            self.last_reported_progress = self.task_progress
+            metadata = {self.key: self.task_progress}
+            self.update_state(meta=metadata)
+
+    @property
+    def task_progress(self):
+        return int(min((100.0 * self.progress / self.total), 100.0))
+
+
+class CeleryTask(Task):
+    """
+    Custom task class so we can add default `after_return` handling which marks the task as
+    succeeded or failed.
+
+    This is set as the Task class on our Celery app, so to track progress on a task, mark it
+    when decorating the task:
+    ```
+        @app.task(bind=True, track_progress=True)
+        def my_task(self):
+            self.progress.increment() # progress tracker
+    ```
+    """
+    # by default, celery does not track task starting itself
+    track_started = True
+    send_events = True
+
+    # custom task option
+    track_progress = False
+
+    _progress_tracker = None
+
+    def after_return(self, status, retval, task_id, args, kwargs, einfo):
+        """
+        Ensures status is updated after task completion, otherwise our signals may not fire
```

Comment:
Celery signals are triggered by state, so this method is defensive to ensure those get fired for the task.